README.md

```diff
@@ -695,200 +775,441 @@ Tuples, lists, and strings are also monoids:
+A paramorphism is a recursive function that has access to the results of the recursive call and the original data structure.
 
 ```python
-# TODO
+def para(numbers, acc=0):
+    if not numbers:
+        return acc
+    head, *tail = numbers
+    return para(tail, acc + head)
+
+# A paramorphism that sums a list of numbers
+assert para([1, 2, 3, 4]) == 10
 ```
 
-### Apomorphism (TODO)
+### Apomorphism
 
-it's the opposite of paramorphism, just as anamorphism is the opposite of catamorphism. Whereas with paramorphism, you combine with access to the accumulator and what has been accumulated, apomorphism lets you `unfold` with the potential to return early.
+An apomorphism is a dual to paramorphism, allowing for early termination in an unfold.
 
-## Setoid (TODO)
+```python
+def apo(predicate, function, seed):
+    result = []
+    while predicate(seed):
+        value, seed, done = function(seed)
+        result.append(value)
+        if done:
+            break
+    return result
+
+# An apomorphism that generates a range of numbers but can stop early
+def range_apomorphism(start, stop, early_stop):
+    return apo(
+        lambda x: x < stop,
+        lambda x: (x, x + 1, x >= early_stop),
+        start
+    )
+
+assert range_apomorphism(0, 10, 5) == [0, 1, 2, 3, 4]
+```
 
-An object that has an `equals` function which can be used to compare other objects of the same type.
+## Setoid
 
-Make array a setoid:
+A setoid is an object that has an `equals` method which can be used to compare other objects of the same type for equality.
 
-```python 
-# TODO
+```python
+class Setoid:
+    def __init__(self, value):
+        self.value = value
+
+    def equals(self, other):
+        return self.value == other.value
+
+# Example usage:
+a = Setoid(1)
+b = Setoid(1)
+c = Setoid(2)
+
+assert a.equals(b)
+assert not a.equals(c)
 ```
 
-## Semigroup (TODO)
+## Semigroup
 
-An object that has a `concat` function that combines it with another object of the same type.
+A semigroup is an algebraic structure with a `concat` method that combines it with another object of the same type.
 
 ```python
-# TODO
+class Semigroup:
+    def __init__(self, value):
+        self.value = value
+
+    def concat(self, other):
+        return Semigroup(self.value + other.value)
+
+# Example usage:
+a = Semigroup("Hello, ")
+b = Semigroup("World!")
+
+assert a.concat(b).value == "Hello, World!"
 ```
 
-## Foldable (TODO)
+## Foldable
 
-An object that has a `reduce` function that applies a function against an accumulator and each element in the array (from left to right) to reduce it to a single value.
+A foldable is an object that has a `reduce` method that applies a function against an accumulator and each element in the object to reduce it to a single value.
 
 ```python
-# TODO
+class Foldable(list):
+    def reduce(self, function, initial):
+        accumulator = initial
+        for value in self:
```

Comment:
Or maybe we don't need `list` base at all: `self.value = [...]` might be enough.